### ISSUE TITLE
pkgman support

### ISSUE THREAD

**Issue by trungnt2910** [trungnt2910/hyclone#4]
While filesystem and network emulation has been implemented in Hyclone, pkgman still requires some magic port to communicate with the kernel.

This magic port's specification is still unknown by Hyclone and undocumented by Haiku.

**Comment by trungnt2910** [comment]
![image](https://user-images.githubusercontent.com/57174311/229338186-729c0ea3-a946-49ca-a827-d44a24183856.png)

We've been able to run `package_daemon`, but `pkgman` is still returning an error.

**Comment by trungnt2910** [comment]
Currently `package_daemon` requires specific details related to mounted volumes.

It would now be necessary to rewrite the whole way HyClone emulates the filesystem. The `vchroot` mechanism that is currently living in `haiku_loader` should be completely re-written in HyClone server instead.

It does not need to be something complicated like `fs_shell`, probably just a function that translates paths (a more robust solution to the current `vchroot`), and statbufs (necessary for us to establish full control over `dev_t`, and probably `uid_t` and `gid_t` in the not so far future.

**Comment by trungnt2910** [comment]
As of 9ab2cf6 the features mentioned above has been implemented.

`package_daemon` is now stuck when trying to communicate with the `packagefs` driver through a magic `ioctl`:

```cpp
		if (ioctl(fd.Get(), PACKAGE_FS_OPERATION_GET_VOLUME_INFO, info,
				bufferSize) != 0) {
			ERROR("Volume::Init(): failed to get volume info: %s\n",
				strerror(errno));
			RETURN_ERROR(errno);
		}
```

This is closely related to #2

**Comment by trungnt2910** [comment]
![image](https://user-images.githubusercontent.com/57174311/230364971-a5b9c03f-6019-4fcf-b76a-840fe0677368.png)

`package_daemon` works!

However, running `pkgman` will immediately crash the `package_daemon` due to a missing `transfer_area` syscall.

**Comment by trungnt2910** [comment]
![image](https://user-images.githubusercontent.com/57174311/230660395-a4774147-db13-430f-815f-805f4c61ba10.png)

As of 70e6ff8, `pkgman search` can run without crashing either itself or `package_daemon`. `transfer_area` also works properly, as shown in the test below (taken from [Haiku's test](https://github.com/haiku/haiku/blob/master/src/tests/system/kernel/transfer_area_test.cpp)).

The error, `No matching packages found.`, is probably due to HyClone not being configured with Haiku's default package sources.

**Comment by trungnt2910** [comment]
Dirty as it may seem, `pkgman` can now conduct package searches on Linux:

![image](https://user-images.githubusercontent.com/57174311/230698124-1b612581-7ae3-4329-b1d4-fd28f0d025e9.png)
![image](https://user-images.githubusercontent.com/57174311/230698133-8d2013b0-04bf-4b61-b3d0-0e54cbfc5ae1.png)

**Comment by trungnt2910** [comment]
![image](https://user-images.githubusercontent.com/57174311/230698789-3c5d1fc2-e630-4c4c-8d62-19d9ff51b7e8.png)

`pkgman install` still needs some more work.

**Comment by trungnt2910** [comment]
![image](https://user-images.githubusercontent.com/57174311/230709665-645ce5dd-14f4-4a7f-ad24-79802fbfdfe2.png)

`pkgman install` is actually _fixing_ broken dependencies on HyClone! WOW!

**Comment by trungnt2910** [comment]
Currently, `pkgman` can resolve dependencies and install packages for us. However:
- All installations require a restart (of HyClone, not of the host). This is because some magic `ioctl`s used to activate and deactivate packages are not implemented in HyClone's `packagefs` driver yet.
- Sometimes downloads may be interrupted by a signal. HyClone uses a realtime signal to send "requests" to `haiku_loader` and `monika` from the server side. This signal has been registered with `SA_RESTART` so it should **not** make syscalls like `read` and `write` fail, but somehow `pkgman` just gets interrupted. There should be some more investigation into this.
- [WSL1 only] WSL1 does not support extended file attributes, therefore lots of (non-fatal) errors will occur during installation as `pkgman` heavily depends on extended attributes in its operation.

**Comment by trungnt2910** [comment]
With extended attributes implemented in 6f657e23a7e81743113e5120a332ee294aee1622 and the rest of the VFS subsystem polished, HyClone is now able to do a `full-sync`:

![image](https://user-images.githubusercontent.com/57174311/230862880-728fff7a-4899-4eee-9f2a-334942d43e99.png)

All the problems mentioned in the above comment has been fixed.

**Comment by trungnt2910** [comment]
![image](https://user-images.githubusercontent.com/57174311/230864586-1f44f576-2555-4790-9a7e-8f93fd82aa05.png)

`pkgman` is also able to do an `uninstall`. I think this is enough to close this issue for now.
